File: java/06_118.java
public class RedBlackTree {
    private static final boolean RED = true;
    private static final boolean BLACK = false;

    private class Node {
        int key;
        Node left, right, parent;
        boolean color; // true for red, false for black

        Node(int key) {
            this.key = key;
            this.color = RED;
            this.left = NIL;
            this.right = NIL;
        }
    }

    private final Node NIL;
    private Node root;

    public RedBlackTree() {
        NIL = new Node(0);
        NIL.color = BLACK;
        root = NIL;
    }

    public void insert(int key) {
        Node node = new Node(key);

        Node y = NIL;
        Node x = root;

        // Find position for new node
        while (x != NIL) {
            y = x;
            if (node.key < x.key) {
                x = x.left;
            } else {
                x = x.right;
            }
        }

        // Set parent of node
        node.parent = y;

        // If tree is empty, make node the root
        if (y == NIL) {
            root = node;
        } else if (node.key < y.key) {
            y.left = node;
        } else {
            y.right = node;
        }

        // Fix the tree
        fixInsert(node);
    }

    private void fixInsert(Node k) {
        Node uncle;

        // While parent is red
        while (k.parent != NIL && k.parent.color == RED) {
            if (k.parent == k.parent.parent.right) {
                uncle = k.parent.parent.left;

                // Case 1: Uncle is red
                if (uncle.color == RED) {
                    uncle.color = BLACK;
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    k = k.parent.parent;
                } else {
                    // Case 2: Uncle is black and k is left child
                    if (k == k.parent.left) {
                        k = k.parent;
                        rightRotate(k);
                    }

                    // Case 3: Uncle is black and k is right child
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    leftRotate(k.parent.parent);
                }
            } else {
                uncle = k.parent.parent.right;

                // Case 1: Uncle is red
                if (uncle.color == RED) {
                    uncle.color = BLACK;
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    k = k.parent.parent;
                } else {
                    // Case 2: Uncle is black and k is right child
                    if (k == k.parent.right) {
                        k = k.parent;
                        leftRotate(k);
                    }

                    // Case 3: Uncle is black and k is left child
                    k.parent.color = BLACK;
                    k.parent.parent.color = RED;
                    rightRotate(k.parent.parent);
                }
            }

            if (k == root) {
                break;
            }
        }
        root.color = BLACK;
    }

    private void leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;

        if (y.left != NIL) {
            y.left.parent = x;
        }

        y.parent = x.parent;

        if (x.parent == NIL) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }

        y.left = x;
        x.parent = y;
    }

    private void rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;

        if (y.right != NIL) {
            y.right.parent = x;
        }

        y.parent = x.parent;

        if (x.parent == NIL) {
            root = y;
        } else if (x == x.parent.right) {
            x.parent.right = y;
        } else {
            x.parent.left = y;
        }

        y.right = x;
        x.parent = y;
    }

    public Node search(int key) {
        return searchHelper(root, key);
    }

    private Node searchHelper(Node node, int key) {
        if (node == NIL || key == node.key) {
            return node;
        }

        if (key < node.key) {
            return searchHelper(node.left, key);
        }
        return searchHelper(node.right, key);
    }

    public void inOrderTraversal() {
        System.out.println("In-order traversal of the Red-Black Tree:");
        inOrderHelper(root);
        System.out.println();
    }

    private void inOrderHelper(Node node) {
        if (node != NIL) {
            inOrderHelper(node.left);
            System.out.print(node.key + "(" + 
                            (node.color ? "RED" : "BLACK") + ") ");
            inOrderHelper(node.right);
        }
    }

    public static void main(String[] args) {
        RedBlackTree rbTree = new RedBlackTree();
        int[] keys = {7, 3, 18, 10, 22, 8, 11, 26};

        for (int key : keys) {
            rbTree.insert(key);
        }

        rbTree.inOrderTraversal();

        int keyToFind = 10;
        Node foundNode = rbTree.search(keyToFind);
        if (foundNode != rbTree.NIL) {
            System.out.println("Found key " + keyToFind + 
                              ", color: " + 
                              (foundNode.color ? "RED" : "BLACK"));
        } else {
            System.out.println("Key " + keyToFind + " not found");
        }
    }
}

File: data structures and algorithms/cs/06_118.cs
using System;
using System.Collections.Generic;

public class RedBlackTree
{
   private static readonly bool RED = true;
   private static readonly bool BLACK = false;

   private class Node
   {
      public int Key;
      public Node Left, Right, Parent;
      public bool Color; // true for red, false for black

      public Node(int key)
      {
         this.Key = key;
         this.Color = RED;
         this.Left = NIL;
         this.Right = NIL;
      }
   }

   private static readonly Node NIL = new Node(0)
   {
      Color = BLACK
   };
   private Node root = NIL;

   public void Insert(int key)
   {
      Node node = new Node(key);

      Node y = NIL;
      Node x = root;

      // Find position for new node
      while (x != NIL)
      {
         y = x;
         if (node.Key < x.Key)
         {
            x = x.Left;
         }
         else
         {
            x = x.Right;
         }
      }

      // Set parent of node
      node.Parent = y;

      // If tree is empty, make node the root
      if (y == NIL)
      {
         root = node;
      }
      else if (node.Key < y.Key)
      {
         y.Left = node;
      }
      else
      {
         y.Right = node;
      }

      // Fix the tree
      FixInsert(node);
   }

   private void FixInsert(Node k)
   {
      Node uncle;

      // While parent is red
      while (k.Parent != NIL && k.Parent.Color == RED)
      {
         if (k.Parent == k.Parent.Parent.Right)
         {
            uncle = k.Parent.Parent.Left;

            // Case 1: Uncle is red
            if (uncle.Color == RED)
            {
               uncle.Color = BLACK;
               k.Parent.Color = BLACK;
               k.Parent.Parent.Color = RED;
               k = k.Parent.Parent;
            }
            else
            {
               // Case 2: Uncle is black and k is left child
               if (k == k.Parent.Left)
               {
                  k = k.Parent;
                  RightRotate(k);
               }

               // Case 3: Uncle is black and k is right child
               k.Parent.Color = BLACK;
               k.Parent.Parent.Color = RED;
               LeftRotate(k.Parent.Parent);
            }
         }
         else
         {
            uncle = k.Parent.Parent.Right;

            // Case 1: Uncle is red
            if (uncle.Color == RED)
            {
               uncle.Color = BLACK;
               k.Parent.Color = BLACK;
               k.Parent.Parent.Color = RED;
               k = k.Parent.Parent;
            }
            else
            {
               // Case 2: Uncle is black and k is right child
               if (k == k.Parent.Right)
               {
                  k = k.Parent;
                  LeftRotate(k);
               }

               // Case 3: Uncle is black and k is left child
               k.Parent.Color = BLACK;
               k.Parent.Parent.Color = RED;
               RightRotate(k.Parent.Parent);
            }
         }

         if (k == root)
         {
            break;
         }
      }
      root.Color = BLACK;
   }

   private void LeftRotate(Node x)
   {
      Node y = x.Right;
      x.Right = y.Left;

      if (y.Left != NIL)
      {
         y.Left.Parent = x;
      }

      y.Parent = x.Parent;

      if (x.Parent == NIL)
      {
         root = y;
      }
      else if (x == x.Parent.Left)
      {
         x.Parent.Left = y;
      }
      else
      {
         x.Parent.Right = y;
      }

      y.Left = x;
      x.Parent = y;
   }

   private void RightRotate(Node x)
   {
      Node y = x.Left;
      x.Left = y.Right;

      if (y.Right != NIL)
      {
         y.Right.Parent = x;
      }

      y.Parent = x.Parent;

      if (x.Parent == NIL)
      {
         root = y;
      }
      else if (x == x.Parent.Right)
      {
         x.Parent.Right = y;
      }
      else
      {
         x.Parent.Left = y;
      }

      y.Right = x;
      x.Parent = y;
   }

   public Node Search(int key)
   {
      return SearchHelper(root, key);
   }

   private Node SearchHelper(Node node, int key)
   {
      if (node == NIL || key == node.Key)
      {
         return node;
      }

      if (key < node.Key)
      {
         return SearchHelper(node.Left, key);
      }
      return SearchHelper(node.Right, key);
   }

   public void InOrderTraversal()
   {
      Console.WriteLine("In-order traversal of the Red-Black Tree:");
      InOrderHelper(root);
      Console.WriteLine();
   }

   private void InOrderHelper(Node node)
   {
      if (node != NIL)
      {
         InOrderHelper(node.Left);
         Console.Write(node.Key + "(" + (node.Color ? "RED" : "BLACK") + ") ");
         InOrderHelper(node.Right);
      }
   }

   public static void Main(string[] args)
   {
      RedBlackTree rbTree = new RedBlackTree();
      int[] keys = { 7, 3, 18, 10, 22, 8, 11, 26 };

      foreach (int key in keys)
      {
         rbTree.Insert(key);
      }

      rbTree.InOrderTraversal();

      int keyToFind = 10;
      Node foundNode = rbTree.Search(keyToFind);
      if (foundNode != NIL)
      {
         Console.WriteLine("Found key " + keyToFind + ", color: " + (foundNode.Color ? "RED" : "BLACK"));
      }
      else
      {
         Console.WriteLine("Key " + keyToFind + " not found");
      }
   }
}